### PR TITLE
test suite

### PR DESCRIPTION
**Pull request by cdmistman**
cucumber was too much - felt like i was debugging the library more than upm itself

this supercedes #124

### CODE REVIEW

File: .semaphore/semaphore.yml [hunk 1]
```diff
@@ -3,121 +3,47 @@ version: v1.0
...
-      prologue:
-        commands:
-          - cache restore nix-installer
-          - sudo /tmp/nix-installer install linux --no-confirm --logger compact
-          - . /nix/var/nix/profiles/default/etc/profile.d/nix-daemon.sh
       jobs:
         - name: nix
           commands:
-            - cache restore nix-store
+            # - cache restore nix-store
```
Comment (cdmistman) [inline .semaphore/semaphore.yml:23]:
```suggestion
```

File: .gitignore [hunk 2]
```diff
@@ -1,6 +1,6 @@
 /cmd/upm/upm
 /dist/
-/internal/statik/
+statik/
```
Comment (cdmistman) [inline .gitignore:3]:
```suggestion
```

File: .semaphore/install_nix.sh [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+# Install Nix. Install the base version if the currently installed version is
+# not what we expect.
+if [ \
+	! -f ~/.nix-profile/etc/profile.d/nix.sh -o \
+	"$(~/.nix-profile/bin/nix --version)" != "nix (Nix) 2.16.1" \
+	]; then
+	curl -L https://releases.nixos.org/nix/nix-2.16.1/install | sh
+fi
+
+. ~/.nix-profile/etc/profile.d/nix.sh
```
Comment (cdmistman) [inline .semaphore/install_nix.sh:18]:
```suggestion
source ~/.nix-profile/etc/profile.d/nix.sh
```

File: nix/upm/default.nix [hunk 4]
```diff
@@ -3,11 +3,16 @@
...
+buildGoModule rec {
   pname = "upm";
   version = rev;
 
-  src = ../../.;
+  src = builtins.path {
+    name = "${pname}-${version}-src";
+    path = ../../.;
+    filter = path: type: type != "directory" ||
+      builtins.baseNameOf path != "test-suite";
```
Comment (cdmistman) [inline nix/upm/default.nix:14]: i'm thinking of changing this to an allowlist instead of a blocklist...

File: test-suite/Add_test.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+		bt.Start(t)
+
+		var pkgs []string
+		switch bt.Backend.Name {
+		default:
+			t.Run(bt.Backend.Name, func(t *testing.T) {
+				t.Skip("no test")
+			})
+			continue
+		}
```
Comment (cdmistman) [inline test-suite/Add_test.go:20]: this pattern will be fleshed out in individual PRs adding per-backend tests, but results in lint failures in this PR :(

File: nix/upm/default.nix (hunk 4 above)
Comment (ryantm) [reply nix/upm/default.nix:14]: Yeah, can it be src and internal?

Comment (cdmistman) [reply nix/upm/default.nix]: done in [6290c1b](https://github.com/replit/upm/pull/126/commits/6290c1b419dc3a46c81272ab171ec3741f631def)

Comment (cdmistman) [reply nix/upm/default.nix]: that didn't actually work, i went back to a blocklist because it's easier and (hopefully) equivalent

**Review by ryantm — APPROVED** [review]
Nice, this looks like a very thorough job of integration testing UPM! Thank you!

**Pull request merged**